Include partition counts in VSM/KT metadata.

cseed opened the pull request:
Use partitionCounts in VSM.count, VSM.toIndexedRowMatrix and ToNormalizedIndexedRowMatrix.

cseed commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -348,6 +348,8 @@ object MatrixIR {
 abstract sealed class MatrixIR extends BaseIR {
   def typ: MatrixType
 
+  def partitionCounts: Option[Array[Long]] = None
```
@tpoterba FYI I added (optional, static) partition counts to MatrixIR and they are used by VSM.count (if available).  They are written out by VSM.write and read back by read.  They should be preserved by map operations but invalided by filter operations, for example, but right now they're only available on MatrixRead.  They should also be computed by the OrderedRVD scan (more PRs coming).

tl;dr: read().count() should now be instantaneous.

tpoterba replied:
🚀

jbloom22 commented on src/main/scala/is/hail/stats/ComputeRRM.scala:
```diff
@@ -97,26 +97,28 @@ object ToNormalizedIndexedRowMatrix {
-    val partitionSizes = vds.rdd2.mapPartitions(it => Iterator.single(it.size)).collect().scanLeft(0)(_ + _)
-    assert(partitionSizes.length == vds.rdd2.getNumPartitions + 1)
-    val pSizeBc = vds.sparkContext.broadcast(partitionSizes)
+    val partitionCounts = vds.partitionCounts()
```
the `// extra leading 0` should be deleted or put above the assert

jbloom22 commented on src/main/scala/is/hail/utils/richUtils/RichRDD.scala:
```diff
@@ -9,6 +9,7 @@ import org.apache.hadoop
 import org.apache.hadoop.io.compress.CompressionCodecFactory
 import org.apache.spark.{NarrowDependency, Partition, TaskContext}
 import org.apache.spark.rdd.RDD
+import org.json4s.jackson.Serialization
```
Serialization is not used

jbloom22 commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -2279,7 +2290,7 @@ class VariantSampleMatrix(val hc: HailContext, val metadata: VSMMetadata,
   }
 
-  def writeMetadata(dirname: String, nPartitions: Int) {
+  def writeMetadata(dirname: String, nPartitions: Int, partitionCounts: Array[Long]) {
```
you can remove nPartitions as a parameter here, and just use partitionCounts.length below

jbloom22 commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -2711,9 +2725,10 @@ class VariantSampleMatrix(val hc: HailContext, val metadata: VSMMetadata,
-    assert(partitionSizes.length == rdd2.getNumPartitions + 1)
-    val pSizeBc = sparkContext.broadcast(partitionSizes)
+    val pCounts = partitionCounts()
+    val partitionStarts = pCounts.scanLeft(0L)(_ + _)
```
I think it's worth moving the partitionStarts def to VSM under partitionCounts, it's already in two places and will be in more.

jbloom22 commented:
Search `.zipWithIndex()` and you'll see five places we used Spark's zipWithIndex, which triggers a job. When partitionCounts are available I think we could avoid that.

cseed commented:
I addressed comments and rebased.

 > Search .zipWithIndex() and you'll see five places we used Spark's zipWithIndex

Good call.  There are additional cases in KT where it should also be used.  I will address these in two additional PRs.
